bundle/sign.go

```diff
@@ -71,3 +83,17 @@ func generatePayload(files []FileInfo, sc *SigningConfig, keyID string) ([]byte,
 	}
 	return json.Marshal(payload)
 }
+
+// GetSigner returns the current Signer
+func GetSigner() Signer {
+	return signer
```

Comment:
Does this not make the signer/verifier global state? Would we ever want to handle more than one signer/verifier? 

I think it would make sense to rather inject/attach the signer/verifier on the bundle or bundle config.

Reply:
This is a good call, and timely, because we're actually about to encounter the need for multiple verifiers in our use case I think.  I'll explore some options for implementing this.  Any suggestions?  Maybe something similar to `runtime.RegisterPlugin`, where a verifier can be registered under a unique key, and bundles can store that key?  And for signers, maybe just a CLI arg on the `build` and `sign` commands?

Reply:
What about this:

configuration
```yaml
bundles:
  my_bundle:
    signing:
      # optional, overrides the plugin id stored in .signatures.json
      plugin:  my_signing_plugin
plugins:
  # unique id for plugin that sets up signer/verifier
  my_signing_plugin:
    # configuration for the plugin
    optional_config_item: foo
```

.signatures.json. Change is backwards compatible. If no plugin exists in json, use default signing implementation, or whatever is explicitly defined in configuration.
```json
{
  "signatures": [ ... ],
  "plugin": "my_signing_plugin"
}
```

`build` and `sign` commands get a new arg. will use existing signing/verification args as well.
```sh
opa build --bundle --plugin my_signing_plugin .
```

Reply:
I guess the only sticking point on that one is that the `sign`/`build` commands won't be able to pass any configuration to the signing plugin, as would happen in the `run` command...

Reply:
Hmm, and this would be the first case where a plugin is used outside the `run` command I think?  The `runtime.RegisterPlugin` and `PluginFactory` wouldn't be in scope for the `sign`/`build` commands, so there would need to be another way to register/get a `Signer`...maybe I'm trying to stretch the plugin implementation too far here?  It works well for Verifier at least.

Maybe we do the `.signatures.json` change, but just tweak the global var for Signer/Verifier to be a collection of Signers/Verifiers and make the Get/Set accept an `id` for the proper implementation?

Reply:
I'll need to give it some thought. Perhaps custom `build`/`sign` applications should be considered external to the OPA binary entirely? I mean, the built-in commands are already more or less just there for the convenience over manually building a tar ball (at least `build`).. so perhaps it would be good enough to just allow extending verification? 🤔  It's pretty late here though so could be I miss something obvious 😅

Reply:
Sounds good! I’ll have a look again tomorrow. Also struck me that we might want AWS KMS signing/verification without having to use a plugin. Is it a convoluted process? I’ve only used their KMS for other purposes before and I remember there were are few quirks to consider.. though it’s been a while so maybe things have changed.

Reply:
We're currently working on an opa-aws-plugin project we plan to opensource eventually, that will bring a slew of AWS-specific functionality to OPA.  I tried to get clarification [in this issue](https://github.com/open-policy-agent/opa/issues/2757) about where I should draw the line between contributing directly to opa or using a separate plugin project, but never got a final recommendation.

I think the actual signing/verification implementation is better suited for a separate plugin project though (similar to the envoy one), because then we can justify including the AWS SDK instead of trying to re-write all that functionality to avoid the extra vendoring.

Using the AWS SDK, it's pretty trivial to implement the signing/verification, with the caveat that we had to duplicate all the JWT stuff happening in opa core (copying the jwx lib and all that).  That could probably be cleaned up to avoid so much duplication.

Reply:
Ok, alternative implementation committed.  Basically, you can register custom verification and signing implementations, there's a new optional `--signing-plugin` arg on `build` and `sign` that tells them which registered impl to use, and the impl used to sign a bundle is recorded in the bundle's `.signatures.json`, so that opa knows which impl to use for verification.  It's up to the custom impls to deal with any configuration needed (instead of trying to shoehorn in the runtime plugins stuff).  Thoughts?